src/actions/etherspotActions.js

```diff
@@ -0,0 +1,106 @@
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+// constants
+import { ACCOUNT_TYPES } from 'constants/accountsConstants';
+import { SET_INITIAL_ASSETS } from 'constants/assetsConstants';
+
+// actions
+import { addAccountAction } from 'actions/accountsActions';
+import { saveDbAction } from 'actions/dbActions';
+
+// services
+import etherspot from 'services/etherspot';
+
+// utils
+import { normalizeWalletAddress } from 'utils/wallet';
+import { reportErrorLog } from 'utils/common';
+
+// types
+import type { Dispatch, GetState } from 'reducers/rootReducer';
+import type SDKWrapper from 'services/api';
+
+
+export const initEtherspotServiceAction = (privateKey: string) => {
+  return async (dispatch: Dispatch, getState: GetState) => {
+    const {
+      session: { data: { isOnline } },
+    } = getState();
+
+    if (!isOnline) return; // nothing to do
+
+    await etherspot.init(privateKey);
+  };
+};
+
+export const importEtherspotAccountsAction = () => {
+  return async (dispatch: Dispatch, getState: GetState, api: SDKWrapper) => {
+    const {
+      session: { data: session },
+      user: { data: user },
+    } = getState();
+
+    if (!session.isOnline) return; // offline, nothing to dp
+
+    if (!etherspot?.sdk) {
+      reportErrorLog('importEtherspotAccountsAction failed: action dispatched when Etherspot SDK was not initialized');
+      return;
+    }
+
+    if (!user) {
+      reportErrorLog('importEtherspotAccountsAction failed: no user');
+      return;
+    }
+
+    const etherspotAccounts = await etherspot.getAccounts();
+    if (!etherspotAccounts) {
+      // Note: there should be always at least one account, it syncs on Etherspot SDK init, otherwise it's failure
+      reportErrorLog('importEtherspotAccountsAction failed: no accounts', { etherspotAccounts });
+      return;
+    }
+
+    // sync accounts with app
+    await Promise.all(etherspotAccounts.map((etherspotAccount) => dispatch(addAccountAction(
+      etherspotAccount.address,
+      ACCOUNT_TYPES.ETHERSPOT_SMART_WALLET,
+      etherspotAccount, // full object as extras
+    ))));
+
+    const { walletId } = user;
+
+    if (!walletId) {
+      reportErrorLog('importEtherspotAccountsAction failed: no walletId', { user });
+      return;
+    }
```

Comment:
As `user` is take from state at the begging of the action method, and redux state should be immutable, then either:
1) This check could be moved just after `if (!user)` check
2) `user` should be re-queried from state, so that it contains updated `walletId`

Reply:
That's fine, we care about the state of user when the action was fired, however, moved the check upper 👍